### ISSUE TITLE
Consumption / Download of Large Files (~5GB) Fails on Rinkeby Marketplace

### ISSUE THREAD

**Issue by kaimeinke** [oceanprotocol/market#301]
Dear DEV-Team,

i am currently testing for an IDO with a new publisher on the Ocean Market. During the initial tests we discovered that large files with a size of approx. 5 GB can not be downloaded / consumed. This is a blocker for the IDO.

To reproduce the issue I created three pools on Rinkeby Ocean Market.

0. https://market.oceanprotocol.com/asset/did:op:e115a919B9eb125BBEa1Ff37c22Ab725B63c1335
1. https://market.oceanprotocol.com/asset/did:op:83c92519f4A791DdF3f8ae0C6912ED3D05953349
2. https://market.oceanprotocol.com/asset/did:op:1044c915af3d0A2456AC3F1F241393ae725D89c9

Within these pools two 5GB test files should be available for consumption. One TXT file and one ZIP file.

The files are currently hosted, just for testing purposes, on my webspace and are 100% available and downloadable via direct links. They were also detected correctly during the publishing process.

0. https://data-brokers.de/wp-content/uploads/2020/11/test5gb.txt
1. https://data-brokers.de/wp-content/uploads/2020/11/test5gb.txt
2. https://data-brokers.de/wp-content/uploads/2020/11/test5gb.zip

I attach what I could see from the developer console. 

Any advice / fix / communication would be highly appreciated as we need to postpone the IDO as it seems if the market does not offer the desired reliability at this moment. We would not be happy to launch this without a successful test.

Thanks a lot and all the best
Kai 

![2020-12-16 11_09_13-](https://user-images.githubusercontent.com/59994024/102336329-51cb0700-3f91-11eb-888b-1b2fe047b7db.png)

![2020-12-16 11_09_48-Explorer](https://user-images.githubusercontent.com/59994024/102336351-58f21500-3f91-11eb-9919-fbb7e6dd056c.png)

![2020-12-16 11_11_29-Einstellungen](https://user-images.githubusercontent.com/59994024/102336364-5d1e3280-3f91-11eb-8d48-86cfff5fad18.png)

**Comment by kaimeinke** [comment]
Sorry to say that I am able to reproduce this issue with a file size of approx 1GB. Now I am afraid if this an issue on the whole marketplace.

4. https://market.oceanprotocol.com/asset/did:op:7979DBAE616fDC75197C994650b801e2B6db4DdB

Corresponding file is: https://data-brokers.de/wp-content/uploads/2020/11/test1gb.zip 

![image](https://user-images.githubusercontent.com/59994024/102357766-fad42a80-3fae-11eb-936a-6ae2bd3732bf.png)

**Comment by kaimeinke** [comment]
Already after the initial creation I have problems with the DDO.

![image](https://user-images.githubusercontent.com/59994024/102358956-771b3d80-3fb0-11eb-9511-bc9308afa3c8.png)

Refreshing multiple times over minutes does no help, the error stays.

Okay, I guess you are on it since nothing loads on Rinkeby Marketplace right now.

![image](https://user-images.githubusercontent.com/59994024/102359396-01fc3800-3fb1-11eb-9dbb-d893e0b1fc0d.png)

**Comment by kaimeinke** [comment]
This is bad. I am able to reproduce this on the mainnet market with a file about 1GB. Please confirm @kremalicious or @trentmc .

This makes my set unsellable and in general larger projects are blocked by this.

![3](https://user-images.githubusercontent.com/59994024/102465451-f1e66600-404d-11eb-8f1f-4b31741540a7.png)

![4](https://user-images.githubusercontent.com/59994024/102465462-f6ab1a00-404d-11eb-89a2-8758c2c0f245.png)

![5](https://user-images.githubusercontent.com/59994024/102465474-fad73780-404d-11eb-80c8-e4be0656eba0.png)

**Comment by kaimeinke** [comment]
Would it be right to assume that the proxy solution fails to deliver the large files due to a size constraint or a timeout configured somehwhere? I really hope this is not something serious to fix.

**Comment by kaimeinke** [comment]
Maybe this helps ...

`GET https://provider.mainnet.oceanprotocol.com/api/v1/services/download?fileIndex=0&documentId=did:op:eF152D1C4F9171E45ceDaf3f06BFB01831246CcE&serviceId=1&serviceType=access&dataToken=0xeF152D1C4F9171E45ceDaf3f06BFB01831246CcE&transferTxId=0x8ba5165fdcddba6995dac369a37cd9c8aa2a9212511657567669dd277e49b38c&consumerAddress=0x01e66950353400E93AEe7F041C0303103E2ef5Ab&signature=0x4f10faf6a0e32502375624e723a71e51670bb462f99a8dab43c52167b1f253a42a2907967d39f43eab67e352024c98be70228fb96b20753b8b4868e08e5b80da1b net::ERR_CONNECTION_CLOSED
(anonymous) @ VM290:1
(anonymous) @ WebServiceConnector.js:289
(anonymous) @ WebServiceConnector.js:129
(anonymous) @ WebServiceConnector.js:59
(anonymous) @ WebServiceConnector.js:31
n @ WebServiceConnector.js:10
e.fetch @ WebServiceConnector.js:283
e.get @ WebServiceConnector.js:179
(anonymous) @ WebServiceConnector.js:225
(anonymous) @ WebServiceConnector.js:129
(anonymous) @ WebServiceConnector.js:59
(anonymous) @ WebServiceConnector.js:31
n @ WebServiceConnector.js:10
e.downloadFile @ WebServiceConnector.js:217
(anonymous) @ Provider.js:417
(anonymous) @ Provider.js:155
(anonymous) @ Provider.js:85
(anonymous) @ Provider.js:57
o @ Provider.js:36
(anonymous) @ Provider.js:398
(anonymous) @ Provider.js:396
(anonymous) @ Provider.js:155
(anonymous) @ Provider.js:85
a @ Provider.js:39
Promise.then (async)
u @ Provider.js:54
a @ Provider.js:39
Promise.then (async)
u @ Provider.js:54
(anonymous) @ Provider.js:57
o @ Provider.js:36
t.download @ Provider.js:377
(anonymous) @ Assets.js:757
(anonymous) @ Assets.js:171
(anonymous) @ Assets.js:101
a @ Assets.js:55
Promise.then (async)
u @ Assets.js:70
(anonymous) @ Assets.js:73
a @ Assets.js:52
t.download @ Assets.js:735
(anonymous) @ index.esm.js:852
s @ index.esm.js:851
(anonymous) @ index.esm.js:882
(anonymous) @ index.esm.js:805
(anonymous) @ index.esm.js:843
(anonymous) @ index.esm.js:819
consume @ index.esm.js:842
(anonymous) @ Consume.tsx:84
c @ runtime.js:63
(anonymous) @ runtime.js:293
(anonymous) @ runtime.js:118
n @ asyncToGenerator.js:3
s @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
q @ Consume.tsx:25
G @ Consume.tsx:25
je @ react-dom.production.min.js:52
Ke @ react-dom.production.min.js:52
(anonymous) @ react-dom.production.min.js:53
Er @ react-dom.production.min.js:100
xr @ react-dom.production.min.js:101
(anonymous) @ react-dom.production.min.js:113
Ie @ react-dom.production.min.js:292
(anonymous) @ react-dom.production.min.js:50
zr @ react-dom.production.min.js:105
Jt @ react-dom.production.min.js:75
Gt @ react-dom.production.min.js:74
t.unstable_runWithPriority @ scheduler.production.min.js:18
Bl @ react-dom.production.min.js:122
Re @ react-dom.production.min.js:292
Yt @ react-dom.production.min.js:73
Show 36 more frames
`
and 

`Logger.js:102 TypeError: Failed to fetch
e.dispatch @ Logger.js:102
e.error @ Logger.js:91
(anonymous) @ index.esm.js:885
Promise.then (async)
(anonymous) @ index.esm.js:811
(anonymous) @ index.esm.js:843
(anonymous) @ index.esm.js:819
consume @ index.esm.js:842
(anonymous) @ Consume.tsx:84
c @ runtime.js:63
(anonymous) @ runtime.js:293
(anonymous) @ runtime.js:118
n @ asyncToGenerator.js:3
s @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
q @ Consume.tsx:25
G @ Consume.tsx:25
je @ react-dom.production.min.js:52
Ke @ react-dom.production.min.js:52
(anonymous) @ react-dom.production.min.js:53
Er @ react-dom.production.min.js:100
xr @ react-dom.production.min.js:101
(anonymous) @ react-dom.production.min.js:113
Ie @ react-dom.production.min.js:292
(anonymous) @ react-dom.production.min.js:50
zr @ react-dom.production.min.js:105
Jt @ react-dom.production.min.js:75
Gt @ react-dom.production.min.js:74
t.unstable_runWithPriority @ scheduler.production.min.js:18
Bl @ react-dom.production.min.js:122
Re @ react-dom.production.min.js:292
Yt @ react-dom.production.min.js:73
`

**Comment by SarahKay99** [comment]
![image](https://user-images.githubusercontent.com/51705696/102482615-96e95900-40b7-11eb-8811-10d6a5eed920.png)

We overcame the CORS error on Kai's machine and were left with a `net::ERR_CONNECTION_CLOSED` error on Rinkeby + Mainnet. 

I was able to reproduce Kai's download problem on the Rinkeby network. I'm pretty sure it is a problem with the server on your end. @kremalicious @trentmc @alexcos20

**Comment by kaimeinke** [comment]
Tried it with Firefox and Metamask on the Mainnet Market to no avail.

```
{
	"GET": {
		"scheme": "https",
		"host": "provider.mainnet.oceanprotocol.com",
		"filename": "/api/v1/services/download",
		"query": {
			"fileIndex": "0",
			"documentId": "did:op:eF152D1C4F9171E45ceDaf3f06BFB01831246CcE",
			"serviceId": "1",
			"serviceType": "access",
			"dataToken": "0xeF152D1C4F9171E45ceDaf3f06BFB01831246CcE",
			"transferTxId": "0x8ba5165fdcddba6995dac369a37cd9c8aa2a9212511657567669dd277e49b38c",
			"consumerAddress": "0x01e66950353400E93AEe7F041C0303103E2ef5Ab",
			"signature": "0x23b6e8fd08e1ec331947b0170fc4c1cbc760e6fe32c5c250ac8f28ac372c388d1d6532d4b313e2715649551f484ea7a4f9382d564b32d60b9b42828ba07b30cf1b"
		},
		"remote": {
			"Adresse": "34.203.60.105:443"
		}
	}
}
```

`error: Consumer signature failed verification: Invalid signature 0x23b6e8fd08e1ec331947b0170fc4c1cbc760e6fe32c5c250ac8f28ac372c388d1d6532d4b313e2715649551f484ea7a4f9382d564b32d60b9b42828ba07b30cf1b for ethereum address 0x01e66950353400E93AEe7F041C0303103E2ef5Ab, documentId did:op:eF152D1C4F9171E45ceDaf3f06BFB01831246CcEand nonce 7.`

```
HTTP/2 401 Unauthorized
server: nginx/1.19.1
date: Thu, 17 Dec 2020 20:52:45 GMT
content-type: application/json
content-length: 339
access-control-allow-origin: https://market.oceanprotocol.com
vary: Origin
strict-transport-security: max-age=15724800; includeSubDomains
X-Firefox-Spdy: h2
```

```
GET /api/v1/services/download?fileIndex=0&documentId=did:op:eF152D1C4F9171E45ceDaf3f06BFB01831246CcE&serviceId=1&serviceType=access&dataToken=0xeF152D1C4F9171E45ceDaf3f06BFB01831246CcE&transferTxId=0x8ba5165fdcddba6995dac369a37cd9c8aa2a9212511657567669dd277e49b38c&consumerAddress=0x01e66950353400E93AEe7F041C0303103E2ef5Ab&signature=0x23b6e8fd08e1ec331947b0170fc4c1cbc760e6fe32c5c250ac8f28ac372c388d1d6532d4b313e2715649551f484ea7a4f9382d564b32d60b9b42828ba07b30cf1b HTTP/2
Host: provider.mainnet.oceanprotocol.com
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:83.0) Gecko/20100101 Firefox/83.0
Accept: */*
Accept-Language: de,en-US;q=0.7,en;q=0.3
Accept-Encoding: gzip, deflate, br
Referer: https://market.oceanprotocol.com/asset/did:op:eF152D1C4F9171E45ceDaf3f06BFB01831246CcE
Content-type: application/json
Origin: https://market.oceanprotocol.com
DNT: 1
Connection: keep-alive
TE: Trailers
```

**Comment by alexcos20** [comment]
this is going deep on the stack, we need to check provider + ocean,js (ocean.py as well could be facing the same problem).
I'm gonna try it with locally deployed components and see how can we improve that

**Comment by kaimeinke** [comment]
Hi @alexcos20 

I created a test with a 5MB file on Rinkeby. 
https://market.oceanprotocol.com/asset/did:op:A5fbc075bC77e11d32b6f7f29843aAF6c520eCD5 

The file source is:
https://data-brokers.de/wp-content/uploads/2020/11/test5mb.zip

Availability can be checked through the sample file download.

A 5MB file will be served without problems.

![image](https://user-images.githubusercontent.com/59994024/102544932-d8790480-40b5-11eb-876d-9b4051681228.png)

**Comment by kaimeinke** [comment]
Next step: 
https://market.oceanprotocol.com/asset/did:op:dd1f3a8B2AE60436d6FBE5d7CC5A3020832Fe527

A 50MB file will be served without problems.

![image](https://user-images.githubusercontent.com/59994024/102545951-3c4ffd00-40b7-11eb-85d5-d16b5690d016.png)

The first failure happens with a file around 500MB.
See: https://market.oceanprotocol.com/asset/did:op:041e64f2fc4DF5a6e36a4BcCBD903DBfDD27398f

A file of 300MB will be served without problems.
https://market.oceanprotocol.com/asset/did:op:7D96201BE92E07AaF040a32F78258DE23FE1049a

A file of 400MB will be served without problems.
https://market.oceanprotocol.com/asset/did:op:b874298930b62EDf50fEa2dBAfedC904739319DB

Now we have a clear range between 400 and 500MB. I will stop testing here as this should have narrowed down the range enough I hope.

**Comment by alexcos20** [comment]
thanks for the info, I have a starting point

**Comment by kaimeinke** [comment]
Hi @alexcos20 , do you use a synchronous or asynchronous proxy? An asynchronous version would be preferable we would guess.

**Comment by alexcos20** [comment]
this is the code in provider: https://github.com/oceanprotocol/provider/blob/70807fbd58722352b44cb0a7c7d98ce10ab08141/ocean_provider/util.py#L41

for the market we are using:
https://github.com/oceanprotocol/ocean.js/blob/93148541fc40a8e2094a27e757e2fce5d72ef3e7/src/ocean/utils/WebServiceConnector.ts#L66  (which uses save if a browser is used - https://www.npmjs.com/package/save-file )

**Comment by alexcos20** [comment]
update: the problem lies in provider, I'm going to update that

**Comment by alexcos20** [comment]
https://github.com/oceanprotocol/provider/issues/40
